File: LeetCode/C/653-tree-两数之和-iv-输入-bst.cpp
/*
 * @lc app=leetcode.cn id=653 lang=cpp
 *
 * [653] 两数之和 IV - 输入 BST
 */
#include "main.h"
// @lc code=start
class Solution {
  public:
    bool findTarget(TreeNode *root, int k) {
        if (root == nullptr)
            return false;
        unordered_set<int> s;
        queue<TreeNode *> q;
        q.push(root);
        while (!q.empty()) {
            auto t = q.front();
            q.pop();
            if (s.insert(k - t->val).second == false)
                return true;
            s.insert(t->val);
            if (t->left)
                q.push(t->left);
            if (t->right)
                q.push(t->right);
        }
        return false;
    }
};
// @lc code=end

File: LeetCode/Go/653-tree-两数之和-iv-输入-bst.go
/*
 * @lc app=leetcode.cn id=653 lang=golang
 *
 * [653] 两数之和 IV - 输入 BST
 */
package leetcode

// @lc code=start
// func findTarget(root *TreeNode, k int) bool {
// 	if root == nil {
// 		return false
// 	}
// 	set := map[int]bool{}
// 	q := []*TreeNode{}
// 	q = append(q, root)
// 	for len(q) != 0 {
// 		t := q[0]
// 		q = q[1:]
// 		if _, ok := set[k-t.Val]; ok {
// 			return true
// 		}
// 		set[t.Val] = true
// 		if t.Left != nil {
// 			q = append(q, t.Left)
// 		}
// 		if t.Right != nil {
// 			q = append(q, t.Right)
// 		}
// 	}
// 	return false
// }
func findTarget(root *TreeNode, k int) bool {
	if root == nil {
		return false
	}
	order := []int{}

	var inorder func(*TreeNode)
	inorder = func(root *TreeNode) {
		if root == nil {
			return
		}
		inorder(root.Left)
		order = append(order, root.Val)
		inorder(root.Right)
	}
	inorder(root)

	l, r := 0, len(order)-1
	for l < r {
		sum := order[l] + order[r]
		if sum == k {
			return true
		} else if sum < k {
			l++
		} else {
			r--
		}
	}
	return false
}

// @lc code=end
